python/ray/serve/pipeline/generate.py

```diff
@@ -0,0 +1,170 @@
+from ray.serve.pipeline.deployment_method_node import DeploymentMethodNode
+from ray.serve.pipeline.deployment_node import DeploymentNode
+
+
+class DeploymentIDGenerator(object):
+    """
+    Generate unique suffix for each given deployment_name requested for id.
+    By default uses deployment_name for the very first time, then append
+    monotonic increasing id to it.
+    """
+
+    __lock = threading.Lock()
+    __shared_state = dict()
+
+    @classmethod
+    def get_deployment_id(cls, deployment_name: str):
+        with cls.__lock:
+            if deployment_name not in cls.__shared_state:
+                cls.__shared_state[deployment_name] = 0
+                return deployment_name
+            else:
+                suffix_num = cls.__shared_state[deployment_name] + 1
+                cls.__shared_state[deployment_name] = suffix_num
+
+                return f"{deployment_name}_{suffix_num}"
+
+def _get_unique_deployment_name(dag_node):
+    """
+    Get unique deployment name for given ClassNode. For the first time we use
+    class name, and subsequent instantition of the same class used in Ray DAG
+    get monotonic increasing id as suffix.
+    """
+    deployment_name = (
+        dag_node.get_options().get("name", None) or dag_node._body.__name__
+    )
+    deployment_name = DeploymentIDGenerator.get_deployment_id(
+        deployment_name
+    )
+    return deployment_name
+
+def _remove_non_default_ray_actor_options(ray_actor_options: Dict[str, Any]):
+    """
+    In Ray DAG building we pass full ray_actor_options regardless if a field
+    was explicitly set. Since some values are invalid, we need to remove them
+    from ray_actor_options.
+    """
+    ray_actor_options = {
+        k: v for k, v in ray_actor_options.items() if v
+    }
+    if ray_actor_options.get("placement_group") == "default":
+        del ray_actor_options["placement_group"]
+    if ray_actor_options.get("placement_group_bundle_index") == -1:
+        del ray_actor_options["placement_group_bundle_index"]
+    if ray_actor_options.get("max_pending_calls") == -1:
+        del ray_actor_options["max_pending_calls"]
+
+    return ray_actor_options
+
+
+def _replace_init_args_with_deployment_handle(args, kwargs):
+    """
+    Deployment can be passed into other DAGNodes as init args. This is supported
+    pattern in ray DAG. Thus we need convert them into deployment handles in
+    ray serve DAG to make end to end DAG executable.
+    """
+    init_args = []
+    init_kwargs = {}
+
+    # TODO: (jiaodong) Need to handle deeply nested deployment in args
+    for arg in args:
+        if isinstance(arg, DeploymentNode):
+            init_args.append(arg._deployment_handle)
+        else:
+            init_args.append(arg)
+    for key, value in kwargs.items():
+        if isinstance(value, DeploymentNode):
+            init_kwargs[key] = value._deployment_handle
+        else:
+            init_kwargs[key] = value
+    return init_args, init_kwargs
```

Comment:
shouldn't we be able to hook into the ray DAG node arg replacement that already handles nested args?

Reply:
this would be a very big limitation, not acceptable for public release IMO

Reply:
Exactly -- that's the most feasible way I can think of as well, pickling / json serialization with hook on `DAGNode` type is the simplest way we handle nested args. 

There're some details to be figured out that I didn't have a passing test yet, since we both 1) want to replace ClassNode with DeploymentNode 2) But for the `deployment` instance we construct and pass into `DeploymentNode`, we need to recursively replace all `DeploymentNode` with handle. I'm thinking maybe a two-pass solution is cleaner.

Let me work on this on branch #3 today and update this to be mergeable.